src/irmin/type/type_pp.ml

```diff
@@ -16,28 +16,78 @@
 
 open Type_core
 
-let t t =
+let t ?(ocaml_syntax = false) t =
   let rec aux : type a. a t -> a pp =
    fun t ppf x ->
     match t with
     | Self s -> aux s.self_fix ppf x
     | Custom c -> c.pp ppf x
     | Map m -> map m ppf x
     | Prim p -> prim p ppf x
-    | _ -> Type_json.pp t ppf x
+    | _ when not ocaml_syntax -> Type_json.pp t ppf x
+    | Var v -> raise (Unbound_type_variable v)
+    | List l -> Fmt.Dump.list (aux l.v) ppf x
+    | Array a -> Fmt.Dump.array (aux a.v) ppf x
+    | Option o -> option o ppf x
+    | Tuple t -> tuple t ppf x
+    | Record r -> record r ppf x
+    | Variant v -> variant v ppf x
   and map : type a b. (a, b) map -> b pp = fun l ppf x -> aux l.x ppf (l.g x)
   and prim : type a. a prim -> a pp =
    fun t ppf x ->
+    match (t, ocaml_syntax) with
+    | Unit, false -> ()
+    | Char, false -> Fmt.char ppf x
+    | Float, false when Float.is_nan x -> Fmt.string ppf "\"nan\""
+    | Float, false when x = infinity -> Fmt.string ppf "\"inf\""
+    | Float, false when x = neg_infinity -> Fmt.string ppf "\"-inf\""
+    | String _, false -> Fmt.string ppf x
+    | Unit, true -> Fmt.string ppf "()"
+    | Char, true -> Fmt.(pf ppf "'%c'" x)
```

Comment:
Let's group `| Foo, false` and `| Foo, true` together in the ordering to keep things lexicographic.